src/std_impls/collections.rs

```diff
@@ -154,64 +95,139 @@ unsafe impl<T: Finalize, S: BuildHasher> Finalize for HashSet<T, S> {
     }
 }
 
-// TUPLES
-unsafe impl<A, B> GcDeref for (A, B)
-where
-    A: GcDeref,
-    B: GcDeref,
-{
-}
-unsafe impl<A, B> GcDrop for (A, B)
-where
-    A: GcDrop,
-    B: GcDrop,
-{
-}
-unsafe impl<A, B> GcSafe for (A, B)
-where
-    A: GcSafe,
-    B: GcSafe,
-{
-}
+/// Vec like structure means that it implemented `Iter<T>`
+#[macro_export]
+macro_rules! sync_vec_like {
+    ($t:ty) => {
+        unsafe impl<T> GcDeref for $t where T: GcDeref {}
+        unsafe impl<T> GcDrop for $t where T: GcDrop {}
+        unsafe impl<T> GcSafe for $t where T: GcSafe {}
+
+        unsafe impl<T: Scan> Scan for $t {
+            #[inline]
+            fn scan(&self, scanner: &mut Scanner<'_>) {
+                for e in self {
+                    scanner.scan(e)
+                }
+            }
+        }
 
-unsafe impl<A: Scan, B: Scan> Scan for (A, B) {
-    #[inline]
-    fn scan(&self, scanner: &mut Scanner<'_>) {
-        self.0.scan(scanner);
-        self.1.scan(scanner);
-    }
-}
+        unsafe impl<T: Finalize> Finalize for $t {
+            unsafe fn finalize(&mut self) {
+                let set = read(self);
+                for mut e in set {
+                    e.finalize();
+                    forget(e);
+                }
+            }
+        }
+    };
+    {$($t:ty,)*} => {
+        $(sync_vec_like!($t);)*
+    };
+    {$($t:ty), *} => {
+        sync_vec_like!($($t,)*);
+    };
+}
+
+sync_vec_like![
+    Vec<T>,
+    VecDeque<T>, // avoid format
+    LinkedList<T>,
+    BTreeSet<T>,
+    BinaryHeap<T>,
+];
+
+/// Map like structure means that it implemented `Iter<K, V>`
+#[macro_export]
```

Comment:
I don't think we want to `macro_export` this. Much better to force either explicit implementation, or using the derive. Otherwise we introducing a macro that can generate potentially unsound code without the word `unsafe` in the name, which is an anti-pattern